doc/i2chipd.1
.\" Automatically generated by Pod::Man 2.28 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "I2CHIPD 1"
.TH I2CHIPD 1 "2017-09-09" "version 20150809" "Raspberry Pi"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
i2chipd \- read values from i2c chips and store data to database
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\&\fBi2chipd\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The \fBi2chipd\fR is used to read different data from i2c chips on Raspberry 
Pi. The data is converted to physical quantity and written to a file in 
directory /var/lib/i2chipd/. The measured values can be optionally inserted 
to a SQLite3 or MySQL database table. The reading interval of each chip is
defined in file /etc/i2chipd_config together with other measurement 
parameters.
.PP
The daemon is started with command
.PP
\&\fBservice\fR \fIi2chipd\fR \fIstart\fR
.PP
and can be stopped any time with
.PP
\&\fBservice\fR \fIi2chipd\fR \fIstop\fR
.PP
The operation of the daemon can be checked from the log file.
.SH "FILES"
.IX Header "FILES"
\&\fI/etc/i2chipd_config\fR             Configuration file.
.PP
\&\fI/etc/init.d/i2chipd\fR             Init script.
.PP
\&\fI/usr/sbin/i2chipd\fR               Daemon code.
.PP
\&\fI/var/lib/i2chipd/bmp180_pressure\fR \s-1BMP180\s0 pressure.
.PP
\&\fI/var/lib/i2chipd/bmp180_temperature\fR \s-1BMP180\s0 temperature.
.PP
\&\fI/var/lib/i2chipd/htu21d_humidity\fR \s-1HTU21D\s0 humidity.
.PP
\&\fI/var/lib/i2chipd/htu21d_temperature\fR \s-1HTU21D\s0 temperature.
.PP
\&\fI/var/lib/i2chipd/mag3110_Bx\fR      \s-1MAG3110\s0 magnetic field Bx.
.PP
\&\fI/var/lib/i2chipd/mag3110_By\fR      \s-1MAG3110\s0 magnetic field By.
.PP
\&\fI/var/lib/i2chipd/mag3110_Bz\fR      \s-1MAG3110\s0 magnetic field Bz.
.PP
\&\fI/var/lib/i2chipd/mag3110_temperature\fR  \s-1MAG3110\s0 temperature.
.PP
\&\fI/var/lib/i2chipd/mpl3115a2_altitude\fR   \s-1MPL3115A2\s0 altitude.
.PP
\&\fI/var/lib/i2chipd/mpl3115a2_pressure\fR   \s-1MPL3115A2\s0 pressure.
.PP
\&\fI/var/lib/i2chipd/mpl3115a2_temperature\fR \s-1MPL3115A2\s0 temperature.
.PP
\&\fI/var/lib/i2chipd/tmp102_temperature\fR  \s-1TMP102\s0 temperature.
.PP
\&\fI/var/run/i2chipd.pid\fR            \s-1PID\s0 file.
.SH "WARNING"
.IX Header "WARNING"
No check is done where the query data is written. Could make some hardware 
unusable.
.SH "AUTHORS"
.IX Header "AUTHORS"
Jaakko Koivuniemi
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIi2cdetect\fR\|(8), \fIi2cset\fR\|(8), \fIi2cget\fR\|(8), \fIsqlite3\fR\|(1), \fImysql\fR\|(1)


src/i2chipd.c
/**************************************************************************
 * 
 * Read chips with I2C and write data to a file and database. 
 *       
 * Copyright (C) 2015 Jaakko Koivuniemi.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 *
 ****************************************************************************
 *
 * Sat Jul  4 11:36:06 CEST 2015
 * Edit: Wed 11 Oct 20:08:24 EDT 2017
 *
 * Jaakko Koivuniemi
 **/

#include "i2chipd.h"
#include "I2cReadBytes.h"
#include "I2cRReadRegBytes.h"
#include "I2cWriteCommand.h"
#include "I2cWriteCommandRead3.h"
#include "InitBMP180.h"
#include "InitHtu21d.h"
#include "InitMag3110.h"
#include "InitMpl3115a2.h"
#include "InitTmp102_1.h"
#include "InitTmp102_2.h"
#include "InitTmp102_3.h"
#include "InitTmp102_4.h"
#include "InsertMySQL.h"
#include "InsertSQLite.h"
#include "ReadBMP180.h"
#include "ReadConfig.h"
#include "ReadHtu21d.h"
#include "ReadMag3110.h"
#include "ReadMpl3115a2.h"
#include "ReadTmp102_1.h"
#include "ReadTmp102_2.h"
#include "ReadTmp102_3.h"
#include "ReadTmp102_4.h"
#include "ReadMySQLTime.h"
#include "ReadSQLiteTime.h"

const int version=20171011; // program version
const char *i2cdev="/dev/i2c-1";
int i2cerr=0; // error flag for i2c communication 
const char *confile="/etc/i2chipd_config";
const char pidfile[200]="/var/run/i2chipd.pid";
int loglev=5; // log level
char message[200]="";

int  dbsqlite=0; 
char *dbfile=0;
int dbmysql=0; 
char *dbhost=0;
char *dbuser=0;
char *dbpswd=0;
char *database=0;
int dbport=3306;

struct bmp180 bmp180c;
struct htu21d htu21dc;
struct mag3110 mag3110c;
struct mpl3115a2 mpl3115a2c;
struct tmp102 tmp102c1, tmp102c2, tmp102c3, tmp102c4;

int cont=1; /* main loop flag */

void stop(int sig)
{
  sprintf(message, "signal %d catched, stop", sig);
  syslog(LOG_NOTICE|LOG_DAEMON, "%s", message);
  cont=0;
}

void terminate(int sig)
{
  sprintf(message, "signal %d catched", sig);
  syslog(LOG_NOTICE|LOG_DAEMON, "%s", message);

  sleep(1);
  syslog(LOG_NOTICE|LOG_DAEMON, "stop");

  cont=0;
}

void hup(int sig)
{
  sprintf(message,"signal %d catched",sig);
  syslog(LOG_NOTICE|LOG_DAEMON, "%s", message);
}


int main()
{  
  int ok=0;

  setlogmask(LOG_UPTO (loglev));
  syslog(LOG_NOTICE|LOG_DAEMON, "i2chipd v. %d started",version); 

  signal(SIGINT,&stop); 
  signal(SIGKILL,&stop); 
  signal(SIGTERM,&terminate); 
  signal(SIGQUIT,&stop); 
  signal(SIGHUP,&hup); 

  dbfile = malloc(sizeof(char)*SQLITEFILENAME_SIZE);
  dbhost = malloc(sizeof(char)*DBHOSTNAME_SIZE);
  dbuser = malloc(sizeof(char)*DBUSER_SIZE);
  dbpswd = malloc(sizeof(char)*DBPSWD_SIZE);
  database = malloc(sizeof(char)*DBNAME_SIZE);

  bmp180c.interval = 0; // disable BMP180 reading
  htu21dc.interval = 0; // disable HTU21D reading
  htu21dc.drop = 0; // drop from loop if i2c error
  mag3110c.interval = 0; // disable MAG3110 reading
  mpl3115a2c.interval = 0; // disable MPL3115A2 reading
  tmp102c1.interval = 0; // disable TMP102 at 0x48 reading 
  tmp102c2.interval = 0; // disable TMP102 at 0x49 reading 
  tmp102c3.interval = 0; // disable TMP102 at 0x4A reading
  tmp102c4.interval = 0; // disable TMP102 at 0x4B reading 

  ReadConfig();

  pid_t pid, sid;
        
  pid=fork();
  if(pid<0) 
  {
    exit(EXIT_FAILURE);
  }

  if(pid>0) 
  {
    exit(EXIT_SUCCESS);
  }

  umask(0);

  /* Create a new SID for the child process */
  sid=setsid();
  if(sid<0) 
  {
    syslog(LOG_ERR|LOG_DAEMON, "failed to create child process"); 
    exit(EXIT_FAILURE);
  }
        
  if((chdir("/")) < 0) 
  {
    syslog(LOG_ERR|LOG_DAEMON, "failed to change to root directory"); 
    exit(EXIT_FAILURE);
  }
        
  /* Close out the standard file descriptors */
  close(STDIN_FILENO);
  close(STDOUT_FILENO);
  close(STDERR_FILENO);

  FILE *pidf;
  pidf=fopen(pidfile,"w");

  if(pidf==NULL)
  {
    sprintf(message,"Could not open PID lock file %s, exiting", pidfile);
    syslog(LOG_ERR|LOG_DAEMON, "%s", message);
    exit(EXIT_FAILURE);
  }

  if(flock(fileno(pidf),LOCK_EX|LOCK_NB)==-1)
  {
    sprintf(message,"Could not lock PID lock file %s, exiting", pidfile);
    syslog(LOG_ERR|LOG_DAEMON, "%s", message);
    exit(EXIT_FAILURE);
  }

  fprintf(pidf,"%d\n",getpid());
  fclose(pidf);

  if( bmp180c.interval > 0 )
  {
    if( InitBMP180() < 0 )
    {
      syslog(LOG_ERR|LOG_DAEMON, "BMP180 calibration data reading failed, drop chip from loop");
      bmp180c.interval = 0;
    }
  }

  if( htu21dc.interval > 0 )
  {
    if( InitHtu21d() < 0 )
    {
      syslog(LOG_ERR|LOG_DAEMON, "HTU21D init failed, drop chip from loop");
      htu21dc.interval = 0;
    }
  }

  if( mag3110c.interval > 0 )
  {
    if( InitMag3110() < 0 )
    {
      syslog(LOG_ERR|LOG_DAEMON, "MAG3110 init failed, drop chip from loop");
      mag3110c.interval = 0;
    }
  }

  if( mpl3115a2c.interval > 0 )
  {
    if( InitMpl3115a2() < 0 )
    {
      syslog(LOG_ERR|LOG_DAEMON, "MPL3115A2 init failed, drop chip from loop");
      mpl3115a2c.interval = 0;
    }
  }

  if( tmp102c1.interval > 0 )
  {
    if( InitTmp102_1() < 0 )
    {
      syslog(LOG_ERR|LOG_DAEMON, "TMP102_1 init failed, drop chip from loop");
      tmp102c1.interval = 0;
    }
  }

  if( tmp102c2.interval > 0 )
  {
    if( InitTmp102_2() < 0 )
    {
      syslog(LOG_ERR|LOG_DAEMON, "TMP102_2 init failed, drop chip from loop");
      tmp102c2.interval = 0;
    }
  }

  if( tmp102c3.interval > 0 )
  {
    if( InitTmp102_3() < 0 )
    {
      syslog(LOG_ERR|LOG_DAEMON, "TMP102_3 init failed, drop chip from loop");
      tmp102c3.interval = 0;
    }
  }

  if( tmp102c4.interval > 0 )
  {
    if( InitTmp102_4() < 0 )
    {
      syslog(LOG_ERR|LOG_DAEMON, "TMP102_4 init failed, drop chip from loop");
      tmp102c4.interval = 0;
    }
  }

  if( dbsqlite == 1 )
  {
    if( ReadSQLiteTime( dbfile ) == 0 ) 
    {
      syslog(LOG_ERR|LOG_DAEMON, "SQLite database read failed, drop database connection");
      dbsqlite=0; 
    }
  }

  if( dbmysql == 1 )
  {
    if( ReadMySQLTime(dbhost, dbuser, dbpswd, database, dbport) == 0 ) 
    {
      syslog(LOG_ERR|LOG_DAEMON, "MySQL database read failed, drop database connection");
      dbmysql=0; 
    }
  }

  int unxs=(int)time(NULL); // unix seconds
  bmp180c.next = unxs; // next time to read BMP180
  htu21dc.next = unxs; // next time to read HTU21D
  mag3110c.next = unxs; // next time to read MAG3110
  mpl3115a2c.next = unxs; // next time to read MPL3115A2
  tmp102c1.next = unxs; // next time to read TMP102 1
  tmp102c2.next = unxs; // next time to read TMP102 2
  tmp102c3.next = unxs; // next time to read TMP102 3
  tmp102c4.next = unxs; // next time to read TMP102 4

  while(cont==1)
  {
    unxs=(int)time(NULL); 

    if( bmp180c.interval > 0 )
    {
      if( ( unxs >= bmp180c.next )||( (bmp180c.next - unxs) > bmp180c.next) ) 
      {
        bmp180c.next = bmp180c.interval + unxs;
        i2cerr = 0;
        ReadBMP180();
        if( i2cerr < 0 )
        {
          syslog(LOG_ERR|LOG_DAEMON, "BMP180 reading failed, drop from loop");
          bmp180c.interval = 0;
        }
      }
    }

    if( htu21dc.interval > 0 )
    {
      if( ( unxs >= htu21dc.next )||( (htu21dc.next - unxs) > htu21dc.next) ) 
      {
        htu21dc.next = htu21dc.interval + unxs;
        i2cerr = 0;
        ReadHtu21d();
        if( i2cerr < 0 )
        {
          if( htu21dc.drop == 1 )
          {
            syslog(LOG_ERR|LOG_DAEMON, "HTU21D reading failed, drop from loop");
            htu21dc.interval = 0;
          }
        }
      }
    }

    if( mag3110c.interval > 0 )
    {
      if( ( unxs >= mag3110c.next )||( (mag3110c.next - unxs) > mag3110c.next) ) 
      {
        mag3110c.next = mag3110c.interval + unxs;
        i2cerr = 0;
        ReadMag3110();
        if( i2cerr < 0 )
        {
          syslog(LOG_ERR|LOG_DAEMON, "MAG3110 reading failed, drop from loop");
          mag3110c.interval = 0;
        }
      }
    }

    if( mpl3115a2c.interval > 0 )
    {
      if( ( unxs >= mpl3115a2c.next )||( (mpl3115a2c.next - unxs) > mpl3115a2c.next) ) 
      {
        mpl3115a2c.next = mpl3115a2c.interval + unxs;
        i2cerr = 0;
        ReadMpl3115a2();
        if( i2cerr < 0 )
        {
          syslog(LOG_ERR|LOG_DAEMON, "MPL3115A2 reading failed, drop from loop");
          mpl3115a2c.interval = 0;
        }
      }
    }

    if( tmp102c1.interval > 0 )
    {
      if( ( unxs >= tmp102c1.next )||( (tmp102c1.next - unxs) > tmp102c1.next) ) 
      {
        tmp102c1.next = tmp102c1.interval + unxs;
        i2cerr = 0;
        ReadTmp102_1();
        if( i2cerr < 0 )
        {
          syslog(LOG_ERR|LOG_DAEMON, "TMP102_1 reading failed, drop from loop");
          tmp102c1.interval = 0;
        }
      }
    }

    if( tmp102c2.interval > 0 )
    {
      if( ( unxs >= tmp102c2.next )||( (tmp102c2.next - unxs) > tmp102c2.next) ) 
      {
        tmp102c2.next = tmp102c2.interval + unxs;
        i2cerr = 0;
        ReadTmp102_2();
        if( i2cerr < 0 )
        {
          syslog(LOG_ERR|LOG_DAEMON, "TMP102_2 reading failed, drop from loop");
          tmp102c2.interval = 0;
        }
      }
    }

    if( tmp102c3.interval > 0 )
    {
      if( ( unxs >= tmp102c3.next )||( (tmp102c3.next - unxs) > tmp102c3.next) ) 
      {
        tmp102c3.next = tmp102c3.interval + unxs;
        i2cerr = 0;
        ReadTmp102_3();
        if( i2cerr < 0 )
        {
          syslog(LOG_ERR|LOG_DAEMON, "TMP102_3 reading failed, drop from loop");
          tmp102c3.interval = 0;
        }
      }
    }

    if( tmp102c4.interval > 0 )
    {
      if( ( unxs >= tmp102c4.next )||( (tmp102c4.next - unxs) > tmp102c4.next) ) 
      {
        tmp102c4.next = tmp102c4.interval + unxs;
        i2cerr = 0;
        ReadTmp102_4();
        if( i2cerr < 0 )
        {
          syslog(LOG_ERR|LOG_DAEMON, "TMP102_4 reading failed, drop from loop");
          tmp102c4.interval = 0;
        }
      }
    }

    sleep(1);
  }

  syslog(LOG_NOTICE|LOG_DAEMON, "remove PID file");
  ok=remove(pidfile);

  return ok;
}
